### PR TITLE
Neon Microblocks

### PR DESCRIPTION
**Pull request by kantai**
Support microblock mining in neon nodes:

* On winning a sortition, a miner will attempt to assemble one microblock (if `mine_microblocks` is set to true)
* Neon miners now will sleep a configurable amount of time before trying to mine a new anchor block (this gives a miner time to receive any microblocks that may be sent from the previous block's winner)

This also adds a new struct to miner.rs, `StacksMicroblockBuilder`, since the microblock assembly has to outlive the StacksBlockBuilder instance. If we want to mine more than _one_ microblock, a single instance of this struct needs to process each of those microblocks (since its ClarityTx isn't committed).

Additionally, this adds some integration testing structure for neon:

```
$ cd testnet
$ BITCOIND_TEST=1 cargo test -- --test-threads 1 --ignored --nocapture neon_integrations
```

With an integration test for mining a microblock.

### CODE REVIEW

File: src/core/mempool.rs [hunk 1]
```diff
@@ -283,6 +285,130 @@ impl MemPoolDB {
...
+        };
+        
+        debug!("Will start scaning mempool at {}/{} height={} ts={}", &next_tip_burn_header_hash, &next_tip_block_hash, next_height, next_timestamp);
+        Ok(Some((next_tip_burn_header_hash, next_tip_block_hash, next_height, next_timestamp)))
+    }
+
+    ///
+    /// Iterate over candidates in the mempool
+    ///  todo may be called multiple times, as the mempool walks back from
+    ///   the provided block.
```
Comment (jcnelson) [inline src/core/mempool.rs:356]: Might want to add a comment here that `todo` will be called once for each bundle of transactions at each ancestor chain tip from the given one, starting with the most recent chain tip and working backwards until there are no more transactions to consider.  Each batch of transactions passed to `todo` will be sorted in fee order, starting with the highest.

File: src/util/secp256k1.rs [hunk 2]
```diff
@@ -61,7 +61,7 @@ pub struct Secp256k1PrivateKey {
     // serde is broken for secp256k1, so do it ourselves
     #[serde(serialize_with = "secp256k1_privkey_serialize", deserialize_with = "secp256k1_privkey_deserialize")]
     key: LibSecp256k1PrivateKey,
-    compress_public: bool
+    pub compress_public: bool
```
Comment (jcnelson) [inline src/util/secp256k1.rs:64]: Could you instead add a `set_compressed()` method to `Secp256k1PrivateKey`, like there is with `Secp256k1PublicKey`?

File: testnet/src/neon_node.rs [hunk 3]
```diff
@@ -515,6 +557,27 @@ impl InitializedNeonNode {
...
+                                mined_block_shh: &BlockHeaderHash,
+                                chain_state: &mut StacksChainState,
+                                keychain: &Keychain,
+                                consumed_execution: ExecutionCost,
+                                mem_pool: &MemPoolDB) -> Result<StacksMicroblock, ChainstateError> {
+        let mut microblock_miner = StacksMicroblockBuilder::new(mined_block_shh.clone(),
+                                                                mined_block_bhh.clone(),
+                                                                chain_state,
+                                                                consumed_execution)?;
+        let mblock_key = keychain.get_microblock_key()
```
Comment (jcnelson) [inline testnet/src/neon_node.rs:570]: Each microblock public key hash must be unique in the chain history.  Not sure this is happening here.

File: testnet/src/tests/neon_integrations.rs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    eprintln!("== REGTEST BLOCK MINED == ");
+    thread::sleep(std::time::Duration::from_millis(gap_ms));
+
+
+    let path = format!("{}/v2/accounts/{}?proof=0",
+                       &http_origin, &spender_addr);
+    let res = client.get(&path).send().unwrap().json::<AccountEntryResponse>().unwrap();
+    eprintln!("{:#?}", res);
+    assert_eq!(u128::from_str_radix(&res.balance[2..], 16).unwrap(), 98300);
+    assert_eq!(res.nonce, 1);
```
Comment (jcnelson) [inline testnet/src/tests/neon_integrations.rs:210]: Can you expand this test so that it mines a sequence of anchored-block/microblock pairs?  i.e. to confirm that the miner can pick up the microblock and build on it?

File: src/chainstate/stacks/miner.rs [hunk 5]
```diff
@@ -56,6 +56,149 @@ use util::vrf::*;
...
+                            txs_to_broadcast.push(mempool_tx.tx);
+                        },
+                        Err(e) => {
+                            match e {
+                                Error::CostOverflowError(cost_before, cost_after, total_budget) => { 
+                                    warn!("Transaction {} reached block cost {}; budget was {}", mempool_tx.tx.txid(), &cost_after, &total_budget);
+                                    clarity_tx.reset_cost(cost_before.clone());
+                                },
+                                _ => {
+                                    warn!("Error processing TX {}: {}", mempool_tx.tx.txid(), e);
```
Comment (jcnelson) [inline src/chainstate/stacks/miner.rs:136]: Should this path include the erred transaction if it was a runtime error, i.e., so the miner gets the fee but doesn't produce an invalid block?

File: src/chainstate/stacks/miner.rs [hunk 6]
```diff
@@ -56,6 +56,149 @@ use util::vrf::*;
...
+            header_reader,
+            prev_microblock_header: None,
+            considered: Some(HashSet::new()),
+        })
+    }
+
+    pub fn mine_next_microblock(&mut self,
+                                mem_pool: &MemPoolDB,
+                                miner_key: &Secp256k1PrivateKey,
+                                miner_pubkey_hash: &Hash160) -> Result<StacksMicroblock, Error> {
```
Comment (jcnelson) [inline src/chainstate/stacks/miner.rs:104]: Should there be a maximum microblock size?  Or do we only care about the size of the entire epoch's data?  I'm leaning more towards the latter, but could be convinced of the former.  Either way, we'll need to track the total epoch size.

File: src/chainstate/stacks/miner.rs [hunk 7]
```diff
@@ -56,6 +56,149 @@ use util::vrf::*;
...
+///       new function. This is required for the `clarity_tx` -- basically, to append transactions
+///       as new microblocks, the builder _needs_ to be able to keep the current clarity_tx "open" 
+pub struct StacksMicroblockBuilder<'a> {
+    anchor_block: BlockHeaderHash,
+    anchor_block_bhh: BurnchainHeaderHash,
+    anchor_block_height: u64,
+    prev_microblock_header: Option<StacksMicroblockHeader>,
+    header_reader: StacksChainState,
+    clarity_tx: Option<ClarityTx<'a>>,
+    considered: Option<HashSet<Txid>>,
```
Comment (jcnelson) [inline src/chainstate/stacks/miner.rs:75]: We will need to track the total epoch's size in this struct.

File: testnet/src/neon_node.rs (hunk 3 above)
Comment (kantai) [reply testnet/src/neon_node.rs:570]: Yep -- the microblock pubkey hash is rotated before building the anchor block (if it rotated during the microblock assembly, it wouldn't match the the key from the anchor block).

File: testnet/src/tests/neon_integrations.rs (hunk 4 above)
Comment (kantai) [reply testnet/src/tests/neon_integrations.rs:210]: That's already what is tested here -- the RPC interface is returning data from the chainstate at the block which _confirms_ the microblock (and for the transaction to have been included, it must have been via a microblock) -- I can add another account info lookup for the miner's nonce (which would test the assertion that blocks are continuing to mine).

File: src/chainstate/stacks/miner.rs (hunk 5 above)
Comment (kantai) [reply src/chainstate/stacks/miner.rs:136]: Probably not -- I think "runtime error handling" needs to be done within `process_transaction` (which handles aborting side-effects for transactions that cannot be included in a block at all), and that any transactions that will deduct fees (and be included in the block) shouldn't return an Err from `process_transaction`. I'm not positive that `process_transaction` is currently handling runtime exceptions correctly, but I think that's unrelated to this PR.

File: src/chainstate/stacks/miner.rs (hunk 6 above)
Comment (kantai) [reply src/chainstate/stacks/miner.rs:104]: Yes -- I think it needs to be included in total epoch size.

File: src/util/secp256k1.rs (hunk 2 above)
Comment (kantai) [reply src/util/secp256k1.rs]: Addressed in d0de5c9cc

File: src/core/mempool.rs (hunk 1 above)
Comment (kantai) [reply src/core/mempool.rs]: Added in d0de5c9cc

**Comment by kantai** [comment]
> I'm not sure if this belongs in this PR, but one challenge with microblocks in general is that the act of accepting an unconfirmed microblock needs to also tentatively update the Stacks chain tip, so that queries on account nonces and balances are accurate (i.e. so I can send multiple transactions from the same account in a microblock stream). This means that the StacksChainState should do a little bit more work than just pre-processing and storing a microblock to its staging DB -- it also needs to build up a (rollback-able) view of the chainstate as if those microblocks were confirmed, and it needs to discard and re-calculate this view as part of accepting an anchored block that confirms them. Moreover, the mempool needs to be updated to accept transactions from the same account on the same anchored block chain tip and with different nonces as long as those transactions are mineable from different microblock chain tips (but, I think this part is easier -- we just make it so transactions attach to both an anchored chain tip and one of its descendant microblocks in the schema).
>
> What do you think? Should this also be done in this PR?

These things definitely should be implemented (and I think there's some nuance to the first part of that -- nodes should be making an informed decision about which unconfirmed microblocks to return data on behalf of), but I think they should be done in a different PR, since they're fairly separable tasks.

File: src/chainstate/stacks/miner.rs [hunk 8]
```diff
@@ -378,7 +521,7 @@ impl StacksBlockBuilder {
     }
```
Comment (lgalabru) [inline src/chainstate/stacks/miner.rs:521]: There might be some cleaning left in this file - I think that initially the function `try_mine_tx` was supposed to handle microblocks.

File: testnet/src/tests/bitcoin_regtest.rs [hunk 9]
```diff
@@ -56,16 +58,34 @@ impl BitcoinCoreController {
...
 
+        eprintln!("bitcoind spawned, waiting for startup");
+        let mut out_reader = BufReader::new(process.stdout.take().unwrap());
+
+        let mut line = String::new();
+        while let Ok(bytes_read) = out_reader.read_line(&mut line) {
+            if bytes_read == 0 {
+                return Err(BitcoinCoreError::SpawnFailed("Bitcoind closed before spawning network".into()))
+            }
+            if line.contains("Done loading") {
```
Comment (lgalabru) [inline testnet/src/tests/bitcoin_regtest.rs:76]: ha, excellent thank you!
Do you mind resolving the todo line 126?

File: src/chainstate/stacks/miner.rs (hunk 8 above)
Comment (kantai) [reply src/chainstate/stacks/miner.rs:521]: Yes -- I think there's a lot in this file that is unused in the stacks node, but removing that and then updating the unit tests appropriately will be a pretty sweeping change. For something like that, which should be more refactoring than anything else, I'd prefer to implement in its own PR -- that way there's not a ton of changes that both change functionality in one place but elsewhere are just refactors.

**Comment by jcnelson** [comment]
> These things definitely should be implemented (and I think there's some nuance to the first part of that -- nodes should be making an informed decision about which unconfirmed microblocks to return data on behalf of), but I think they should be done in a different PR, since they're fairly separable tasks.

Sure, this can be a separate PR.  Regarding tracking unconfirmed microblocks, what do you think about just making it the caller's problem by requiring the caller to pass a particular block sequence number (or microblock confirmation depth) form which to service the particular query?

File: src/chainstate/stacks/miner.rs (hunk 7 above)
Comment (kantai) [reply src/chainstate/stacks/miner.rs:75]: Updated in 808abde

File: src/chainstate/stacks/miner.rs (hunk 6 above)
Comment (kantai) [reply src/chainstate/stacks/miner.rs:106]: Updated in 808abde

File: testnet/stacks-node/src/tests/bitcoin_regtest.rs [hunk 10]
```diff
@@ -34,7 +35,7 @@ impl BitcoinCoreController {
         
         let mut command = Command::new("bitcoind");
         command
-            // .stdout(Stdio::piped())
+            .stdout(Stdio::piped())
             .arg("-conf=/dev/null") // todo(ludo): nix only
```
Comment (jcnelson) [inline testnet/stacks-node/src/tests/bitcoin_regtest.rs:39]: This option shouldn't be necessary

File: testnet/stacks-node/src/tests/neon_integrations.rs [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+#[ignore]
+fn microblock_integration_test() {
+    if env::var("BITCOIND_TEST") != Ok("1".into()) {
+        return
+    }
+
+    // used to specify how long to wait in between blocks.
+    //   we could _probably_ add a hook to the neon node that
+    //   would remove some of the need for this
+    let gap_ms = 5_000;
```
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:108]: How hard would this be?  I'm not comfortable with non-deterministic testing if it can be avoided.  Can we do something like poll the neon node's `/v2/info` endpoint to see how far it is in processing the burnchain?  Can we add an API endpoint for getting the current Stacks chain tip and its height, so we can poll on that too?

File: testnet/stacks-node/src/tests/neon_integrations.rs [hunk 12]
```diff
@@ -0,0 +1,212 @@
+use super::{make_stacks_transfer_mblock_only, SK_1, ADDR_4, to_addr};
```
Comment (jcnelson) [inline testnet/stacks-node/src/tests/neon_integrations.rs:1]: Can you add some tests to verify that we can't exceed the epoch block compute or size budgets when mining?

File: testnet/stacks-node/src/tests/neon_integrations.rs (hunk 11 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/neon_integrations.rs:108]: Depends on how it is implemented. Unfortunately, just polling `/v2/info/` wouldn't work, because the thing we want to wait on is the node finishing all of the processing for a new burnchain block (e.g., if it can mine a block, that it has done so and broadcasted it's transaction, or if it's just trying to register a VRF key, that it has done that). I'll add a (test-only) Mutex guarded field to the neon node for 'blocks_processed', which could be monitored rather than relying on the gap.

Comment (jcnelson) [reply testnet/stacks-node/src/tests/neon_integrations.rs:108]: Right -- we need a way to poll on the Stacks chain tip.  This could be part of `/v2/info`, but it currently isn't now.

Comment (kantai) [reply testnet/stacks-node/src/tests/neon_integrations.rs:108]: I don't think even that would be enough here -- we're trying to wait for the node to have finished broadcasting it's block commit, which wouldn't be reflected in the chain state until the next sortition.

Comment (kantai) [reply testnet/stacks-node/src/tests/neon_integrations.rs:129]: Added in 0dfa4cc1f

File: testnet/stacks-node/src/tests/neon_integrations.rs (hunk 12 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/neon_integrations.rs]: Yep -- added a test in 0dfa4cc1f

File: testnet/stacks-node/src/tests/bitcoin_regtest.rs (hunk 10 above)
Comment (kantai) [reply testnet/stacks-node/src/tests/bitcoin_regtest.rs]: Yep -- removed in 0dfa4cc1f

File: testnet/stacks-node/src/tests/neon_integrations.rs (hunk 11 above)
Comment (jcnelson) [reply testnet/stacks-node/src/tests/neon_integrations.rs:129]: Thanks! Looks much better

File: src/core/mempool.rs [hunk 13]
```diff
@@ -284,6 +286,134 @@ impl MemPoolDB {
...
+            debug!("Done scanning mempool -- no chain tips at height {}", next_height);
+            return Ok(None);
+        }
+        
+        let ancestor_tip = {
+            let mut headers_tx = chainstate.headers_tx_begin()?;
+            // `next_height` is 1-indexed, since stacks blocks start at height 1.
+            // The headers DB, however, is 0-indexed, since it includes the genesis
+            // block header hash.  Account for this.
+            let next_header_height = next_height.saturating_sub(1);
```
Comment (jcnelson) [inline src/core/mempool.rs:311]: Just a heads-up -- in #1563, we eliminate `next_header_height` altogether, since the headers DB now starts at 0 (as it should have all along).  I fixed this bug in my tree, but you may need to do it here as well if #1563 merges first.

**Comment by jcnelson** [comment]
@lgalabru do you have any thoughts on this PR?

**Comment by lgalabru** [comment]
@jcnelson apologies, I will allocate some time today.

File: testnet/stacks-node/src/run_loop/neon.rs [hunk 14]
```diff
@@ -10,6 +10,14 @@ use stacks::burnchains::bitcoin::{BitcoinNetworkType,
 use super::RunLoopCallbacks;
 
 /// Coordinating a node running in neon mode.
+#[cfg(test)]
+pub struct RunLoop {
+    config: Config,
+    pub callbacks: RunLoopCallbacks,
```
Comment (lgalabru) [inline testnet/stacks-node/src/run_loop/neon.rs:16]: Not part of this PR, but what’s our take on reintroducing these callbacks on Neon? They were pretty handy for orchestrating tests.

File: src/core/mempool.rs [hunk 15]
```diff
@@ -286,6 +288,129 @@ impl MemPoolDB {
         })
     }
 
+    fn walk(&self, chainstate: &mut StacksChainState, tip_burn_header_hash: &BurnchainHeaderHash, tip_block_hash: &BlockHeaderHash, tip_height: u64) -> Result<Option<(BurnchainHeaderHash, BlockHeaderHash, u64, u64)>, ChainstateError> {
+        // Walk back to the next-highest
+        // ancestor of this tip, and see if we can include anything from there.
+        let next_height = MemPoolDB::get_previous_block_height(&self.db, tip_height)?.unwrap_or(0);
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/core/mempool.rs:294]: Not exactly -- it's the next parent that has transactions in the mempool.

File: testnet/stacks-node/src/run_loop/neon.rs [hunk 16]
```diff
@@ -18,13 +26,41 @@ pub struct RunLoop {
...
+            callbacks: RunLoopCallbacks::new(),
         }
     }
 
+    #[cfg(test)]
+    pub fn new(config: Config) -> Self {
+        Self {
+            config,
+            callbacks: RunLoopCallbacks::new(),
+            blocks_processed: std::sync::Arc::new(std::sync::atomic::AtomicU64::new(0)),
```
[the remark this answers is not in the document]
Comment (kantai) [reply testnet/stacks-node/src/run_loop/neon.rs:42]: The advantage of not doing an import here is that these types only appear in #[cfg(test)]

File: testnet/stacks-node/src/run_loop/neon.rs (hunk 14 above)
Comment (kantai) [reply testnet/stacks-node/src/run_loop/neon.rs:16]: I'd be supportive of it, though I think the callbacks may end up having slightly different timing in Neon

File: testnet/stacks-node/src/tests/neon_integrations.rs [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+    }
+}
+
+fn wait_for_runloop(blocks_processed: &Arc<AtomicU64>) {
+    let start = std::time::Instant::now();
+    while blocks_processed.load(Ordering::SeqCst) == 0 {
+        if start.elapsed() > std::time::Duration::from_secs(PANIC_TIMEOUT_SECS) {
+            panic!("Timed out waiting for run loop to start");
+        }
+        thread::sleep(std::time::Duration::from_millis(100));
```
[the remark this answers is not in the document]
Comment (kantai) [reply testnet/stacks-node/src/tests/neon_integrations.rs:58]: Yep -- fixed in f6dcdd83e

File: testnet/stacks-node/src/tests/bitcoin_regtest.rs [hunk 18]
```diff
@@ -103,7 +122,6 @@ fn bitcoind_integration_test() {
     let mut run_loop = RunLoop::new(conf);
 
     run_loop.callbacks.on_burn_chain_initialized(|burnchain_controller| {
-        // todo(ludo): we need to wait for bitcoind to be ready.
         sleep_ms(5000);
```
[the remark this answers is not in the document]
Comment (kantai) [reply testnet/stacks-node/src/tests/bitcoin_regtest.rs]: Removed in f6dcdd83e

File: src/core/mempool.rs [hunk 19]
```diff
@@ -286,6 +288,129 @@ impl MemPoolDB {
...
+        let mut next_tips = MemPoolDB::get_chain_tips_at_height(&self.db, next_height)?;
+        if next_tips.len() == 0 {
+            // we're done -- no more chain tips
+            debug!("Done scanning mempool -- no chain tips at height {}", next_height);
+            return Ok(None);
+        }
+        
+        let ancestor_tip = {
+            let mut headers_tx = chainstate.headers_tx_begin()?;
+            match StacksChainState::get_index_tip_ancestor(&mut headers_tx, &StacksBlockHeader::make_index_block_hash(tip_burn_header_hash, tip_block_hash), next_height)? {
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/core/mempool.rs]: Yep, linted in f6dcdd83e

File: src/chainstate/stacks/miner.rs [hunk 20]
```diff
@@ -56,6 +56,158 @@ use util::vrf::*;
...
+        let result = mem_pool.iterate_candidates(
+            &self.anchor_block_bhh, &self.anchor_block, self.anchor_block_height, &mut self.header_reader,
+            |micro_txs| {
+                for mempool_tx in micro_txs.into_iter() {
+                    if considered.contains(&mempool_tx.metadata.txid) {
+                        continue;
+                    } else {
+                        considered.insert(mempool_tx.metadata.txid.clone());
+                    }
+                    if mempool_tx.tx.anchor_mode != TransactionAnchorMode::OffChainOnly && mempool_tx.tx.anchor_mode != TransactionAnchorMode::Any {
```
[the remark this answers is not in the document]
Comment (kantai) [reply src/chainstate/stacks/miner.rs]: Good idea -- moved in f6dcdd83e

**Comment by kantai** [comment]
> Quick additional question, are the events emitted by transactions processed in microblocks handled by the event dispatcher?

Yep -- they'll be returned when the relayer processes the microblocks (which happens when the next anchor block commits to them).

**Comment by kantai** [comment]
Merging!
